docs: Add Web3 gateway installation guide (#250)
oasisprotocol/docs: docs/node/web3.md

---
description: Web3 gateway for Emerald and Sapphire ParaTimes
---

# Oasis Web3 Gateway for your EVM ParaTime

This guide will show you how to set up the Oasis Web3 gateway for

> EVM-compatible ParaTimes such as Emerald and Sapphire.

Reviewer: ```suggestion
EVM-compatible ParaTimes, such as Emerald and Sapphire.
```

:::caution

Each ParaTime requires its own instance of the Web3 gateway!

:::

## Prerequisites

### Hardware

* CPU:
  * Minimum: 2.0 GHz x86-64 CPU
  * Recommended: 2.0 GHz x86-64 CPU with 2 cores/vCPUs
* Memory:
  * Minimum: 4 GB of ECC RAM
  * Recommended: 8 GB of ECC RAM
* Storage:
  * Minimum: 300 GB of SSD or NVMe fast storage
  * Recommended: 500 GB of SSD or NVMe fast storage

:::info

To put the figures above into perspective, the Web3 gateway for Emerald with
PostgreSQL encountered **210 GB** of database growth in ~5 months between
Nov 18, 2021 and Apr 11, 2022 (since the [Emerald Mainnet launch]).

:::

### PostgreSQL

The Web3 gateway stores blockchain data in a
[PostgreSQL](https://www.postgresql.org/) database version **13.3** or higher.

> Install it by following the instructions specific to your operating system. You
> will need to create a separate database for each ParaTime. We also suggest that
> you create and use a different username for connecting to the database for each
> ParaTime.

Reviewer: ```suggestion
Install it by following the instructions specific to your operating system.

Because each ParaTime requires its own instance of the Web3 gateway, you will have to create a separate database and a separate user for each Web3 gateway instance.
```

Lets promote good practices.

### Oasis ParaTime Client Node

The Web3 gateway requires a locally deployed ParaTime-enabled Oasis Node.
First, follow the
[Oasis ParaTime Client Node](run-your-node/paratime-client-node.mdx)
guide on how to configure the Oasis client node with one or more ParaTimes.

> Make sure you check the network parameters page ([Mainnet], [Testnet]) and use

Reviewer: In other docs we only link to Mainnet network parameters ([example](https://github.com/oasisprotocol/docs/pull/221#discussion_r968187689)). Or should we start linking to both?

Author: In developer docs we usually refer to both due to obvious reasons ([1](https://docs.oasis.io/dapp/emerald/writing-dapps-on-emerald#testnet-and-mainnet), [2](https://docs.oasis.io/dapp/emerald/integrating-band-oracle-smart-contract#deploy-oracle), [3](https://docs.oasis.io/dapp/sapphire/guide#testnet-and-mainnet), [4](https://docs.oasis.io/dapp/emerald/)). I think writing (Mainnet, Testnet) in parentheses is a good practice.

Reviewer: Agreed. When we come to it, we should adopt this practice everywhere.

the exact combination of the Oasis node/ParaTime versions!

Next, we suggest that you slightly modify the ParaTime section of your Oasis
node configuration. By default, the Emerald and Sapphire ParaTimes prevent
you from executing *expensive* transactions such as simulating EVM calls
without paying for gas. If your Oasis node instance will only be used by the
Web3 gateway(s) and otherwise not exposed to the internet, we suggest that you
enable expensive transactions by adding the following to your Oasis node
config:

```yaml title='config.yml'
# ... sections not relevant are omitted ...
runtime:
  mode: client
  paths:
    - <emerald_bundle_path>
    - <sapphire_bundle_path>

  config:
    # Use the following IDs:
    # 000000000000000000000000000000000000000000000000e2eaa99fc008f87f # Emerald on Mainnet
    # 00000000000000000000000000000000000000000000000072c8215e60d5bca7 # Emerald on Testnet
    # 000000000000000000000000000000000000000000000000a6d1e3ebf60dff6c # Sapphire on Testnet

>     "<emerald_paratime_id>":

Reviewer: ```suggestion
    "{{ emerald_paratime_id }}":
```
Elsewhere we use the Jinja2/Mustache-type of variables (e.g. https://docs.oasis.io/node/run-your-node/paratime-client-node#configuration).

      estimate_gas_by_simulating_contracts: true
      allowed_queries:
        - all_expensive: true

>     "<sapphire_paratime_id>":

Reviewer: ```suggestion
    "{{ sapphire_paratime_id }}":
```

      estimate_gas_by_simulating_contracts: true
      allowed_queries:
        - all_expensive: true
```

> 

Reviewer: ```suggestion

Before using this configuration you should collect the following information to replace the `{{ ... }}` variables present in the configuration file:

* `{{ emerald_paratime_id }}`: The Emerald ParaTime ID from the network parameters.
  * Emerald on Mainnet: `000000000000000000000000000000000000000000000000e2eaa99fc008f87f`
  * Emerald on Testnet: `00000000000000000000000000000000000000000000000072c8215e60d5bca7`
* `{{ sapphire_paratime_id }}`: The Sapphire ParaTime ID from the network parameters.
  * Sapphire on Testnet: `000000000000000000000000000000000000000000000000a6d1e3ebf60dff6c`

```
Or should we let developers look them up under network parameters?

Author: No, let's define them here for convenience. Alternatively we could maybe define some global docusaurus variables for stuff like discord link, public web3 endpoints, paratime ids etc. so we have them defined in one place.

## Download the Web3 Gateway

Check the required version of the Web3 gateway based on which network you will
be using it with ([Mainnet], [Testnet]). Next, download the Oasis-provided

> binaries from the [official github repository][github-releases].

Reviewer: ```suggestion
binaries from the [official GitHub repository][github-releases].
```

Author: Good catch.

Alternatively, you can download the source release and compile it yourself.

> Consult the [README] file for more information.

Reviewer: ```suggestion
Consult the [README.md] file for more information.
```
Because you are referring to a file, it might make sense to specify the exact filename (or even relative path)?

> ## Running the Gateway on Testnet/Mainnet

Reviewer: ```suggestion
## Running the Web3 Gateway on Testnet/Mainnet
```

> Copy the content below to the config file of your gateway and adapt it

Reviewer: ```suggestion
Copy the content below to the config file of your Web3 gateway and adapt it
```

accordingly to fit your needs:

```yaml title='gateway.yml'
# Uncomment the runtime_id below.
#runtime_id: "000000000000000000000000000000000000000000000000e2eaa99fc008f87f" # Emerald on Mainnet
#runtime_id: "00000000000000000000000000000000000000000000000072c8215e60d5bca7" # Emerald on Testnet
#runtime_id: "000000000000000000000000000000000000000000000000a6d1e3ebf60dff6c" # Sapphire on Testnet
# Path to your internal.sock file in the root Oasis node datadir.

> node_address: "unix:<path-to-oasis-node-unix-socket>"

Reviewer: This requires running services in Docker containers or under the same Linux user (see https://github.com/oasisprotocol/oasis-core/issues/4850). There is also a debug parameter that can set up the Unix socket to be outside of the root Oasis node datadir `debug.grpc.internal.socket_path`.

> node_address: "unix:<path-to-oasis-node-unix-socket>"

Reviewer: ```suggestion
node_address: "unix:{{ oasis_node_unix_socket }}"
```
Elsewhere we use the Jinja2/Mustache-type of variables (e.g. https://docs.oasis.io/node/run-your-node/paratime-client-node#configuration).

# By default, we index the entire blockchain history.
# If you are low on disk space or you use the gateway just for submitting transactions, enable
# pruning below.
enable_pruning: false
pruning_step: 100000
indexing_start: 0

log:
  level: debug
  format: json

database:
  # Change host and port, if PostgreSQL is running somewhere else.
  host: "127.0.0.1"
  port: 5432
  # Enter your database name, username and password.

>   db: <postgresql_db>
>   user: <postgresql_user>
>   password: <postgresql_password>

Reviewer: ```suggestion
  db: {{ postgresql_db }}
  user: {{ postgresql_user }}
  password: {{ postgresql_password }}
```

  dial_timeout: 5
  read_timeout: 10
  write_timeout: 5
  max_open_conns: 0

gateway:
  # Uncomment the chain_id below.

>   #chain_id: 42262 # Emerald on Mainnet
>   #chain_id: 42261 # Emerald on Testnet
>   #chain_id: 23295 # Sapphire on Testnet

Reviewer: ```suggestion
  chain_id: {{ chain_id }}
```
And in the comment below the example values takes from network parameters pages.

  http:
    # Change host to your external IP address if you have users accessing Web3 from the outside.
    host: "localhost"
    port: 8545
    cors: ["*"]
  ws:
    # Change host to your external IP address if you have users accessing Web3 from the outside.
    host: "localhost"
    port: 8546
    origins: ["*"]
  method_limits:
    get_logs_max_rounds: 100
```

:::tip

All configuration settings can also be set via environment variables. For
example, you can set the database password by exporting:

```shell
DATABASE__PASSWORD=<postgresql_password>
```

environment variable.

:::

> Start the gateway:

Reviewer: We should emphasize that this command is just for testing. In production environments, a service manager should be set up (like https://docs.oasis.io/node/run-your-node/prerequisites/system-configuration#non-root).

guide on how to configure the Oasis client node with one or more ParaTimes.
Always use the exact combination of the Oasis node/ParaTime versions as
published on the network parameters page for either [Mainnet] or [Testnet]. 

Apart from the transactions that happen on-chain and produce some effects,
there are also a number of read-only queries implemented in the Oasis protocol
and EVM. Some of them may be quite resource-hungry such as simulating EVM calls
and are disabled by default to avoid DDOS attacks. If your Oasis node instance
will only be used by you and your Web3 gateway(s), you can safely enable
*expensive* transactions by adding the following to your Oasis node config:

```yaml title='config.yml'
# ... sections not relevant are omitted ...
runtime:
  mode: client
  paths:
    - {{ emerald_bundle_path }}
    - {{ sapphire_bundle_path }}

  config:
    "{{ emerald_paratime_id }}":
      estimate_gas_by_simulating_contracts: true
      allowed_queries:
        - all_expensive: true
    "{{ sapphire_paratime_id }}":
      estimate_gas_by_simulating_contracts: true
      allowed_queries:
        - all_expensive: true
```

> In the config above replace `{{ ... }}` placeholders with actual Paratime IDs:

Reviewer: ```suggestion
In the config above replace `{{ ... }}` placeholders with actual ParaTime IDs:
```

* `{{ emerald_paratime_id }}`:

>   * Emerald on Mainnet: `000000000000000000000000000000000000000000000000e2eaa99fc008f87f`
>   * Emerald on Testnet: `00000000000000000000000000000000000000000000000072c8215e60d5bca7`
> * `{{ sapphire_paratime_id }}`:
>   * Sapphire on Testnet: `000000000000000000000000000000000000000000000000a6d1e3ebf60dff6c`

Reviewer: ```suggestion
  * Emerald on [Mainnet]: `000000000000000000000000000000000000000000000000e2eaa99fc008f87f`
  * Emerald on [Testnet]: `00000000000000000000000000000000000000000000000072c8215e60d5bca7`
* `{{ sapphire_paratime_id }}`:
  * Sapphire on [Testnet]: `000000000000000000000000000000000000000000000000a6d1e3ebf60dff6c`
```
What do you think about linking to network parameters pages in this way?

Author: Yeah, good idea.

    port: 8545
    cors: ["*"]
  ws:
    # Change host to your external IP address if you have users accessing Web3 from the outside.
    host: "localhost"
    # Use different port for each Web3 gateway instance, if all run locally.
    port: 8546
    origins: ["*"]
  method_limits:
    get_logs_max_rounds: 100
```

Use the following placeholder values:

- `{{ paratime_id }}`: The ID of the Emerald or Sapphire ParaTime which you are
  configuring the Web3 gateway for (see [above](#oasis-paratime-client-node)).
- `{{ oasis_node_unix_socket }}`: Path to the `internal.sock` file created by
  the Oasis node.
- `{{ postgresql_db }}`, `{{ postgresql_user }}`, `{{ postgresql_password }}`:
  Database name and credentials for your PostgreSQL database.
- `{{ chain_id }}`: The chain ID of the EVM network:
  - Emerald on Mainnet: `42262`
  - Emerald on Testnet: `42261`
  - Sapphire on Testnet: `23295`

:::tip

All configuration settings can also be set via environment variables. For
example, instead of setting up the database password in the config file above
you can export:

```shell
DATABASE__PASSWORD=your_password_here
```

:::

> To start the gateway invoke:

Reviewer: ```suggestion
To start the Web3 gateway invoke:
```
